Handle escape directive

schribl opened the pull request:
This change scans the first line if it is a comment for an escape directive as described in the docker [Documentation](https://docs.docker.com/engine/reference/builder/#escape). In case this is set the regex for the line continuation and the newly introduced state for the line continuation character is set. In addition the `_rstrip_backslash` was renamed to `_rstrip_eol` and takes one additional argument for the line continuation character.

Fixes #109 

# Maintainers will complete the following section

- [x] Commit messages are descriptive enough
- [x] Code coverage from testing does not decrease and new code is covered

MartinBasti commented:
/packit copr-build

MartinBasti commented:
Can you please update commit message with better description please? (That one you wrote to PR for example).

Otherwise looks great!

MartinBasti commented on dockerfile_parse/parser.py:
```diff
@@ -262,9 +262,11 @@ def _clean_comment_line(line):
         insnre = re.compile(r'^\s*(\S+)\s+(.*)$')  # matched group is insn
         contre = re.compile(r'^.*\\\s*$')          # line continues?
         commentre = re.compile(r'^\s*#')           # line is a comment?
+        directivere = re.compile(r'^\s*#\s*escape\s*=\s*(\\|`)\s*$', re.I)  # escape directive
```
Dockerfile supports more (2 for now) directives, can we name it `escape_directive_re`?

MartinBasti commented on dockerfile_parse/parser.py:
```diff
@@ -275,6 +277,13 @@ def _clean_comment_line(line):
             # as a multi-line instruction can be interjected with comments.
             if commentre.match(line):
+                # once support for python versions before 3.8 is dropped use walrus operator
+                if lineno == 0 and directivere.match(line):
```
If I read docs correctly, `lineno == 0` is not always true. Directive can be on next lines if all previous lines are comments.

MartinBasti replied:
Ahh sorry, I read it incorrectly

schribl replied:
I think you are right here, the current implementation will fail if there is first a `syntax` directive and in the next line a `escape` directive. But the docs support this case

MartinBasti commented:
Uff there is an old pytest on RHEL7
```
============================= test session starts ==============================
platform linux2 -- Python 2.7.5 -- py-1.4.32 -- pytest-2.7.0 -- /usr/bin/python
rootdir: /builddir/build/BUILD/dockerfile-parse-1.0.0, inifile: 
collecting ... collected 0 items / 1 errors

==================================== ERRORS ====================================
____________________ ERROR collecting tests/test_parser.py _____________________
tests/test_parser.py:33: in <module>
    class TestDockerfileParser(object):
tests/test_parser.py:1356: in TestDockerfileParser
    pytest.param('\\', '`', marks=pytest.mark.xfail),
E   AttributeError: 'module' object has no attribute 'param'
=========================== 1 error in 0.22 seconds ============================
error: Bad exit status from /var/tmp/rpm-tmp.946Bu7 (%check)
```

schribl commented:
I pushed some changes:
* updated the logic to be more in line with the docker spec (syntax directive before escape directive)
* added a few more tests
* updated tests to work with pytest 2.7.x and latest pytest (they have incompatibles for `xfail` in `parameterize` so it now looks ugly)

MartinBasti commented:
/packit copr-build

MartinBasti reviewed (approved):
:rocket: Thank you!

ben-alkov reviewed (approved):
LGMT, thanks!
